Introduce Contribution Doc (#151)
operator-framework/operator-controller: CONTRIBUTING.md

Reviewer: Do we want to include:

1. The build process?
2. The test process?
3. Required tools for building and testing (e.g. `kind`).

Reviewer: If we use taskfiles (https://github.com/operator-framework/enhancements/issues/126), this will potentially become important. Alternatively, if this ought to be part of the README, then I'd reference that earlier.

Reviewer 2: The README should be the first thing anyone will see when they get to the repo, so it makes more sense to me to have setup, build and test should be called out either there.

# How to Contribute

Operator Controller is an Apache 2.0 licensed project and accepts contributions via GitHub pull requests (PRs).

> Thanks for you interest in contributing to the Operator-Controller. 

Reviewer 3: Minor grammatical/phrasing change:
```suggestion
Thank you for your interest in contributing to the Operator-Controller. 
```

Author: Addressed in the latest commit.

As you may or may not know, the Operator-Controller project aims to deliver the user experience described in the [Operator Lifecycle Manager (OLM) V1 Product Requirements Document (PRD)](https://docs.google.com/document/d/1-vsZ2dAODNfoHb7Nf0fbYeKDF7DUqEzS9HqgeMCvbDs/edit). The design requirements captured in the OLM V1 PRD were born from customer and community feedback based on the experience they had with the released version of [OLM V0](github.com/operator-framework/operator-lifecycle-manager).

The user experience captured in the OLM V1 PRD introduces many requirements that are best satisfied by a microservices architecture. The OLM V1 experience currently relies on four projects:
- [The Operator-Controller project](https://github.com/operator-framework/operator-controller/), which is the top level component allowing users to specify operators they'd like to install.
- [The Catalogd project](https://github.com/operator-framework/catalogd/), which hosts operator content and helps users discover installable content.

> - [The Deppy project](https://github.com/operator-framework/deppy/), which enables the operator-controller to identify valid installs and upgrades from the list of installable content provided via the Catalod project

Reviewer 3: Tiny fix and ending period for consistency:
```suggestion
- [The Deppy project](https://github.com/operator-framework/deppy/), which enables the operator-controller to identify valid installs and upgrades from the list of installable content provided via the Catalogd project.
```

Author: Addressed in the latest commit.

- [The RukPak project](https://github.com/operator-framework/rukpak/), which facilitates the installation of operators.

Each of the projects listed above have their own governance, release milestones, and release cadence. However, from a technical perspective, the "OLM V1 experience" matches the experienced offered by the operator-controller project, the top level component which introduces dependencies on RukPak, Deppy, and the Catalogd projects.

## How do we collaborate

> "We need to accept that random issues and pull requests will show up" - Joe L.

> Before diving into our process for coordinating community efforts, I think it's important to set the expectation that Open Source development can be messy. Any effort to introduce a formal workflow for project contributions will almost certainly be circumvented by new community users. Rather than pestering users to subscribe to a project-specific process, we strive to make it as simply as possible to provide valuable feedback. With that in mind, changes to the project will almost certainly follow this process:

Author: Addressed in the latest commit.

> 1. The community engages in discussion in the #olm-dev slack channel.

Author: Addressed in the latest commit.

> 2. The community creates GitHub Issues, GitHub Discussions, or pull requests in the appropriate repos based on (1) and potentially continue the discussion.

Reviewer 3: Mostly a preference nit:
```suggestion
2. The community creates GitHub Issues, GitHub Discussions, or pull requests in the appropriate repos based on (1) to continue the discussion.
```

Author: Addressed in the latest commit.

> 3. The community utilize the Working Group and talk about (1) and (2) stuff and anything else that comes to mind.

Author: Addressed in the latest commit.

## Where are Operator Controller Milestones?

Ongoing or previous Operator-Controller milestones can always be found in the [milestone section of our GitHub Repo](https://github.com/operator-framework/operator-controller/milestones).

### How are Subproject Issues Tracked?

As discussed earlier, the operator-controller adheres to a microservice architecture, where multiple projects contribute to the overall experience. As such, when designing an operator-controller milestone, the community may need to file an issue against RukPak, Deppy, or Catalogd. Unfortunately, the operator-controller milestone cannot contain issues from one of its subprojects. As such, we've introduced the concept of a "Dependency Issue", described below:

> Dependency Issues: An issue tracked in a milestone that "points" to an issue in another project with a URL.

## Submitting Issues

Unsure where to submit an issue? 
- The Operator-Controller project
- The RukPak project
- The Deppy project

> - The Catalogd project

Author: Good question, I added similar links and descriptions earlier in the doc, perhaps making that a section and linking it here is the right idea?

> - The Catalogd project

Reviewer 4: Ah I see, I don't think it hurts to be extra verbose in case that section is missed or someone is going through and looking at all the links and forgets that information was already provided. I think there is quite a bit of information to digest when it comes to OLMv1 so being that bit more verbose and having *some* repetition might be beneficial

> - The Catalogd project

Author: Addressed in the latest commit.

### Code Review

Contributing PRs with a reasonable title and description can go a long way with helping the PR through the review
process.

When opening PRs that are in a rough draft or WIP state, prefix the PR description with `WIP: ...` or create a draft PR.
This can help save reviewer's time by communicating the state of a PR ahead of time. Draft/WIP PRs can be a good way to
get early feedback from reviewers on the implementation, focusing less on smaller details, and more on the general
approach of changes.

When contributing changes that require a new dependency, check whether it's feasible to directly vendor that
code [without introducing a new dependency](https://go-proverbs.github.io/).

Currently, PRs require at least one approval from a operator-controller maintainer in order to get merged.

### Code style

The coding style suggested by the Golang community is used throughout the operator-controller project:

- [CodeReviewComments](https://github.com/golang/go/wiki/CodeReviewComments)
- [EffectiveGo](https://golang.org/doc/effective_go)

In addition to the linked style documentation, operator-controller formats Golang packages using the `golangci-lint` tool. Before

> submitting a PR, please run `make lint` locally and commit the results. This will help expedite the review process,

Reviewer 3: I don't think this make target is actually in the project yet :sweat_smile:

Author: Oh, let's add that as a follow up issue.

# How to contribute

Operator Controller is an Apache 2.0 licensed project and accepts contributions via GitHub pull requests (PRs).

> This document outlines some conventions on commit message formatting, contact points for developers, and other resources

Author: I've created [an issue](https://github.com/operator-framework/operator-controller/issues/160) to add some Architecture docs

to help Operator Controller contributors.

## Communication

- Email: [operator-framework-olm-dev](mailto:operator-framework-olm-dev@googlegroups.com)
- Slack: [#olm-dev](https://kubernetes.slack.com/archives/C0181L6JYQ2)
- Google Group: [olm-gg](https://groups.google.com/g/operator-framework-olm-dev)
- Weekly in Person Working Group Meeting: [olm-wg](https://github.com/operator-framework/community#operator-lifecycle-manager-working-group) 

## Contribution flow

> Any new contribution should be accompanied by a new or existing issue. This issue can help track work, discuss the

Reviewer: Although we mention trivial fixes don't need a separate issue; where's the dividing line? A PR itself could be a bug report and fix in one, and having to create an issue (which, unrelatedly, share the same number spaces as PRs) seems like duplication of effort.

approach of changes.

When contributing changes that require a new dependency, check whether it's feasible to directly vendor that
code [without introducing a new dependency](https://go-proverbs.github.io/).

Currently, PRs require at least one approval from a operator-controller maintainer in order to get merged.

### Code style

The coding style suggested by the Golang community is used throughout the operator-controller project:

- [CodeReviewComments](https://github.com/golang/go/wiki/CodeReviewComments)
- [EffectiveGo](https://golang.org/doc/effective_go)

In addition to the linked style documentation, operator-controller formats Golang packages using the `golangci-lint` tool. Before
submitting a PR, please run `make lint` locally and commit the results. This will help expedite the review process,
focusing less on style conflicts, and more on the design and implementation details.

Please follow this style to make the operator-controller project easier to review, maintain and develop.

### Documentation

If the contribution changes the existing APIs or user interface it must include sufficient documentation to explain the
new or updated features.

> The Operator Controller documentation is primarily housed at the root-level README.

Reviewer: Link? (we have links to all other stuff, perhaps here too?)

Author: Linked

# How to contribute

Operator Controller is an Apache 2.0 licensed project and accepts contributions via GitHub pull requests (PRs).

This document outlines some conventions on commit message formatting, contact points for developers, and other resources

> to help Operator Controller contributors.

Reviewer 2: Do we require contributors to agree to the CNCF contributor license agreement?

Author: Right now we don't enforce it, could be a good thing to bring up at the WG meeting.

## Communication

- Email: [operator-framework-olm-dev](mailto:operator-framework-olm-dev@googlegroups.com)
- Slack: [#olm-dev](https://kubernetes.slack.com/archives/C0181L6JYQ2)
- Google Group: [olm-gg](https://groups.google.com/g/operator-framework-olm-dev)

> - Weekly in Person Working Group Meeting: [olm-wg](https://github.com/operator-framework/community#operator-lifecycle-manager-working-group) 

Reviewer 5: +1 ! Nice addition to have the details handy :)

## Getting started

- Fork the repository on GitHub.
- Clone onto your local development machine via `git clone https://github.com/{$GH-USERNAME}/operator-controller`.

> - Add the operator-framework/operator-controller as an upstream.

Reviewer: Git hates trailing whitespace

> - Add the operator-framework/operator-controller as an upstream.

Author: Added the spaces as I would prefer a newline.

  remote: `git remote add upstream https://github.com/operator-framework/operator-controller`.

> - Create a new branch from the default `main` branch and begin development.

Reviewer 2: There is a more detailed getting started section in README.md, so we may want to consider moving this section there.

Author: Done.

## Contribution flow

Any new contribution should be accompanied by a new or existing issue. This issue can help track work, discuss the
design and implementation, and help avoid wasted efforts of multiple people working on the same issue. Trivial changes,
like fixing a typo in the documentation, do not require the creation of a new issue.

> 

Reviewer 2: Should we have some guidelines on how to create and tag issues for features, bugs or security vulnerabilities?

Author: Might be a better use of our time to create a few [issue templates](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/configuring-issue-templates-for-your-repository).

Author: Captured this idea in an issue: https://github.com/operator-framework/operator-controller/issues/154

### Small Contributions

For simple contributions, this is a rough outline of what a contributor's workflow looks like:

- Identify or create a GitHub Issue.
- Create a topic branch from where to base the contribution. This is usually the main branch.
- Make commits of logical units.

> - Make sure commit messages are in the proper format (see below).

Reviewer 2: should this link to a subsection?

Author: Great idea!

> - Push changes in a topic branch to a personal fork of the repository.

Reviewer 2: We possibly need to mention test/coverage expectations with codecov.

Author: I don't think that we have any hardcore code coverage expectations, but I'll call out the need for tests.

- Submit a pull request to the operator-framework/operator-controller repository.

> - Wait and respond to feedback from the maintainers listed in the OWNERS file.

Reviewer 2: Outlining the expected response from the maintainers might be useful, so contributors know what to expect.

Author: Good idea, I'll add a section asking them to tag us if we don't respond within a week.

reach out to the community by commenting on the issue or by messaging us on the [#olm-dev](https://kubernetes.slack.com/archives/C0181L6JYQ2) Slack Channel.

If you are interested in proposing a "focus" for a milestone, the contribution workflow should be similar to the following:

- Create a [GitHub Issue](https://github.com/operator-framework/operator-controller/issues/new), labeled with "milestone-proposal", that includes a link to a [HackMD](https://hackmd.io) capturing the proposed changes.
- Add an item to the [OLM Working Group and Issue Triage Meeting Agenda](https://docs.google.com/document/d/1Zuv-BoNFSwj10_zXPfaS9LWUQUCak2c8l48d0-AhpBw/edit) to introduce your design to the community.
- Introduce your proposed changes at the weekly [OLM Working Group Meeting](https://github.com/operator-framework/community#operator-lifecycle-manager-working-group) 

The community will then discuss the proposal, consider the benefits and cost of the feature, iterate on the design, and decide if it project should pursue the design. If accepted, the design will be queued up for a subsequent milestone.

Thanks for contributing!

> ### Code Review

Reviewer 4: Something that might be nice to add here is some guidelines on when and how to reach out to maintainers directly about PRs. For example, is it expected that my PR may take 2 weeks to get a review? If I haven't received a review in a week, where & who should I reach out to to make sure my changes are getting reviewed?

Author: Good point! We may want to add an OWNERS or MAINTAINERS file and then simply mention to tag the individuals in that group.

Author: I've created [an issue](https://github.com/operator-framework/operator-controller/issues/153) to track this work

Contributing PRs with a reasonable title and description can go a long way with helping the PR through the review
process.

When opening PRs that are in a rough draft or WIP state, prefix the PR description with `WIP: ...` or create a draft PR.
This can help save reviewer's time by communicating the state of a PR ahead of time. Draft/WIP PRs can be a good way to
get early feedback from reviewers on the implementation, focusing less on smaller details, and more on the general
approach of changes.

When contributing changes that require a new dependency, check whether it's feasible to directly vendor that
code [without introducing a new dependency](https://go-proverbs.github.io/).

> Currently, PRs required at least one approval from a RukPak maintainer in order to get merged.

Reviewer 4: Do PRs actually require reviews from another subproject's maintainers right now?
```suggestion
Currently, PRs required at least one approval from a operator-controller maintainer in order to get merged.
```

Author: Ah, I had started this file by copying RukPak's CONTRIBUTION.md, I had meant to swap all RukPak references with operator-controller. I will address this in all places.

> Currently, PRs required at least one approval from a operator-controller maintainer in order to get merged.

Reviewer 3: Very small fix:
```suggestion
Currently, PRs require at least one approval from a operator-controller maintainer in order to get merged.
```

Author: Addressed.

> Currently, PRs required at least one approval from a operator-controller maintainer in order to get merged.

Reviewer 5: Can we add a section mentioning requirements to open a PR. Something like [this](https://sdk.operatorframework.io/docs/contribution-guidelines/opening-pull-requests/#pr-checklist). Since this repository is in a fairly new state, we could ask for PR description, tests, relevant documentation as some tips for opening a PR.

Reviewer: That's info that could be added to the PR template.

Author: I think this is covered by [the existing PR template](https://github.com/operator-framework/operator-controller/blob/main/.github/workflows/pull_request_template.md).

Author: @varshaprasad96 please mark this as resolved if you agree.

### Code style

> The coding style suggested by the Golang community is used throughout the RukPak project:

Reviewer 4: ```suggestion
The coding style suggested by the Golang community is used throughout the operator-controller project:
```

- [CodeReviewComments](https://github.com/golang/go/wiki/CodeReviewComments)
- [EffectiveGo](https://golang.org/doc/effective_go)

> In addition to the linked style documentation, RukPak formats Golang packages using the `golangci-lint` tool. Before

Reviewer 4: ```suggestion
In addition to the linked style documentation, operator-controller formats Golang packages using the `golangci-lint` tool. Before
```

submitting a PR, please run `make lint` locally and commit the results. This will help expedite the review process,
focusing less on style conflicts, and more on the design and implementation details.

> Please follow this style to make the RukPak project easier to review, maintain and develop.

Reviewer 4: ```suggestion
Please follow this style to make the operator-controller project easier to review, maintain and develop.
```
